Bootstrap study (#5)
jackbsteinberg/std-toast: study-group/Toast-Implementation-Studies.md

- [Material-UI Snackbar](#material-ui-snackbar)
- [Android `Toast`](#android-toast)
- [react-toastify](#react-toastify)

> - [ngx-toastr](#ngx-toast)

Reviewer: Was this change intended?

Author: Probably not, I think that came from some of the merge shenanigans
```suggestion
- [ngx-toastr](#ngx-toastr)
```

- Comes with default icons, can change by modifying CSS.
- **Note**: On mobile there's currently no simple way to stop the timeout, accomplished by hover on desktop.

## [Bootstrap Toast](https://getbootstrap.com/docs/4.2/components/toasts/)
The Bootstrap Toast is a component from the Bootstrap library to provide a lightweight, customizable notification in the library's distinct style.

### Design & Code Samples
The Toast component, like many others in the Bootstrap library, is put directly into HTML:

```html
<div class="toast" role="alert" aria-live="assertive" aria-atomic="true">
  <div class="toast-header">
    <img src="..." class="rounded mr-2" alt="...">
    <strong class="mr-auto">A Bootstrap Toast</strong>
    <button type="button" class="ml-2 mb-1 close" data-dismiss="toast" aria-label="Close">
      <span aria-hidden="true">&times;</span>
    </button>
  </div>
  <div class="toast-body">
    Hello, world! This is a Bootstrap Toast!
  </div>
</div>
```

> Bootstrap uses classes to denote special elements of the components it constructs, like the classes `toast`, `toast-header`, and `toast-body`.

Reviewer 2: this file does not seem to be following semantic line-breaking.

Author: My bad, it's still been a bit strange to get used to, I'll go through and change that everywhere I can find

The example uses many `aria` attributes, indicating a deliberate effort for accessibility.

> Multiple Toasts can be wrarpped together in a `div` to stack intuitively.

Reviewer 2: ```suggestion
Multiple Toasts can be wrapped together in a `div` to stack intuitively.
```

> Multiple Toasts can be wrarpped together in a `div` to stack intuitively.

Reviewer 2: I'm not sure what "stack intuitively" means. Might be good to include images and now that I think of it, images in some of the other studies could be helpful too (I don't have specific examples but maybe having a snapshot of any interesting examples would be good. Nobody is likely to actually make one of these examples run to see what it looks like.

Author: That's a good idea, something I had discussed with Domenic was creating a file of informative snapshots to go along with the study and link relevant ones to sections in this doc. Do you think I should do that, or just add in a few to this doc?

Author: @fergald I'm going to make a branch once the content is all pushed in to clean it up, by editing and adding some relevant images.

Reviewer 2: Yeah, it's not urgent but I think surveys of UI components will benefit from images (but this being mostly an API survey, not so much, mostly for demonstrating results of non-obvious APIs).

      <div class="toast-header">
        <img src="..." class="rounded mr-2" alt="...">
        <strong class="mr-auto">Bootstrap</strong>
        <small class="text-muted">just now</small>
        <button type="button" class="ml-2 mb-1 close" data-dismiss="toast" aria-label="Close">
          <span aria-hidden="true">&times;</span>
        </button>
      </div>
      <div class="toast-body">
        See? Just like this.
      </div>
    </div>

    <div class="toast" role="alert" aria-live="assertive" aria-atomic="true">
      <div class="toast-header">
        <img src="..." class="rounded mr-2" alt="...">
        <strong class="mr-auto">Bootstrap</strong>
        <small class="text-muted">2 seconds ago</small>
        <button type="button" class="ml-2 mb-1 close" data-dismiss="toast" aria-label="Close">
          <span aria-hidden="true">&times;</span>
        </button>
      </div>
      <div class="toast-body">
        Heads up, Toasts will stack automatically
      </div>
    </div>
  </div>
</div>
```

Using native styling with the `position: absolute;` attribute on, the Toasts can be intuitively positioned on the screen.

> Note that the order the Toasts display is determined by their DOM order, not the order in which they're displayed.

Reviewer: This sentence sounds a little funny, although I know what you mean. Could it be rephrased?

Author: Yeah I'll try and rephrase it

Author: ```suggestion
Note that the order the Toasts display in is determined by their DOM order, _not_ the order in which they are triggered.
```

To display, select all with the `.toast` class and initialize, then select one by its element ID to display:

```js
// Initializes all Toast elements
$('.toast').toast({
  animation: true, // adds a CSS fade
  autohide: true, // automatically closes the Toast
  delay: 5000 // sets time to wait before closing the Toast
})

$('#single-toast-id').toast('show')
```

### Notable Features & Details

#### Pure HTML View
Bootstrap Toasts are purely HTML elements highly styled with CSS, and are easily customizable.

#### Events
The library offers events for `show`, `shown`, `hide`, and `hidden`, to represent the moments before and after the Toast is shown / hidden.

#### Accessibility
The documentation is explicit about suggesting wrapping the toasts in `aria-live`, to announce the content via screen reader without moving focus.
It also recommends including `aria-atomic="true"` to ensure the Toast is read as a single atomic element, 

> and live changes don't disrupt the screen-reader user.

Reviewer: This is great info!

Author: Thanks 👍

#### Z-Index
Bootstrap components that are meant to be overlayed use an arbitrary z-index value, in the range of ~1000.

### Takeaways

- Low JavaScript configuration, high HTML customization.
- Uses the same `.toast(options)` method for `show`, `hide`, and `dispose`, by passing in any of those strings as `options`.
- The developer needs to be cognizant of DOM order for multiple toasts.
  -  Calling `show` out order would still display Toasts in DOM order, a potentially unexpected behavior.
- Requires `util.js` if building the JavaScript from source.

> - When disabling `autohide` the developer must provide a close button, it will not be added by default.

Reviewer: ```suggestion
- When disabling `autohide` the developer must provide a close button; it will not be added by default.
```
